subsys/shell/shell_serial_loader.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2021 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <sys/byteorder.h>
+#include <drivers/uart.h>
+#include <shell/shell.h>
+#include <shell/shell_uart.h>
+
+static inline bool is_ascii(uint8_t data)
+{
+	if ((data >= 0x30 && data <= 0x39) || (data >= 0x61 && data <= 0x66) ||
+	    (data >= 0x41 && data <= 0x46)) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+volatile unsigned char *bytes;
+volatile uint32_t *data;
+int sum;
+int chunk_element;
+char chunk[2];
+bool littleendian;
+
+#define CHAR_CAN 0x18
+#define CHAR_DC1 0x11
+
+static void bypass_cb(const struct shell *sh, uint8_t *recv, size_t len)
+{
+	bool escape = false;
+	static uint8_t tail;
+	uint8_t byte;
+
+	if (tail ==  CHAR_CAN && recv[0] == CHAR_DC1) {
+		escape = true;
+	} else {
+		for (int i = 0; i < (len - 1); i++) {
+			if (recv[i] == CHAR_CAN && recv[i + 1] == CHAR_DC1) {
+				escape = true;
+				break;
+			}
+		}
+	}
+
+	if (escape) {
+		shell_print(sh, "Number of bytes read: %d", sum);
+		shell_set_bypass(sh, NULL);
+
+		if (!littleendian) {
+			while (sum > 4) {
+				*data = __bswap_32(*data);
+				data++;
+				sum = sum - 4;
+			}
+			if (sum % 4 == 0) {
+				*data = __bswap_32(*data);
+			} else if (sum % 4 == 2) {
+				*data = __bswap_16(*data);
+			} else if (sum % 4 == 3) {
+				*data = __bswap_24(*data);
+			}
+		}
+		return;
+	}
+
+	tail = recv[len - 1];
+
+	if (is_ascii(*recv)) {
+		chunk[chunk_element] = *recv;
+		chunk_element++;
+	}
+
+	if (chunk_element == 2) {
+		byte = (uint8_t)strtoul(chunk, NULL, 16);
+		*bytes = byte;
+		bytes++;
+		sum++;
+		chunk_element = 0;
+	}
+}
+
+static int cmd_load(const struct shell *sh, size_t argc, char **argv)
+{
+	littleendian = false;
+	char *arg;
+
+	chunk_element = 0;
+	sum = 0;
+
+	while (argc >= 2) {
+		arg = argv[1] + (!strncmp(argv[1], "--", 2) && argv[1][2]);
+		if (!strncmp(arg, "-e", 2))
+			littleendian = true;
+		else if (!strcmp(arg, "--")) {
+			argv++;
+			argc--;
+			break;
+		} else if (arg[0] == '-' && arg[1])
+			shell_print(sh, "Unknown option \"%s\"", arg);
+		else
+			break;
+		argv++;
+		argc--;
+	}
+
+	bytes = (unsigned char *)strtol(argv[1], NULL, 0);
+	data = (uint32_t *)strtol(argv[1], NULL, 0);
+
+	shell_set_bypass(sh, bypass_cb);
+	shell_print(sh, "Loading...\npress ctrl-x ctrl-q to escape");
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_serial,
+			       SHELL_CMD_ARG(load, NULL,
+					     "Usage:\n"
+					     "serial load [options] [address]\n"
+					     "Options:\n"
+					     "-e\tlittle-endian parse",
+					     cmd_load, 2, 1),
+			       SHELL_SUBCMD_SET_END);
+
+SHELL_CMD_REGISTER(serial, &sub_serial, "serial commands", NULL);
```

Comment:
i wonder if `serial` as a root name is not too generic? Don't have better proposal now but if it refer to serial/uart then it is no longer valid since command will be supported on any transport.

Reply:
That's a good point. I think standalone `load` might not be enough either though. Would `shell load` work? It's always used from the shell command anyway so it would always use the same transport. Or maybe it should start with `mem` or `memory`? Is there any other way to read/write the memory from the shell in Zephyr where this could fit?

Reply:
maybe something like `mem` could be good, it could be later extended with `write`, `read`, etc.

Reply:
Can you then rename and move the file to `subsys/shell/modules/memory_service.c`

Reply:
#36666 there will be a good place to add it. `devmem load`

Reply:
Moved it to devmem, thanks for the suggestion!